internal/encryption/keyprovider/openbao/config.go

```diff
@@ -0,0 +1,104 @@
+package openbao
+
+import (
+	"errors"
+	"fmt"
+
+	openbao "github.com/openbao/openbao/api"
+	"github.com/opentofu/opentofu/internal/encryption/keyprovider"
+)
+
+type Config struct {
+	Address string `hcl:"address,optional"`
+	Token   string `hcl:"token,optional"`
+
+	KeyName   string        `hcl:"key_name"`
+	KeyLength DataKeyLength `hcl:"key_length,optional"`
+}
+
+func (c Config) Build() (keyprovider.KeyProvider, keyprovider.KeyMeta, error) {
+	if c.KeyName == "" {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Message: "no key name found",
+		}
+	}
+
+	if c.KeyLength == 0 {
+		c.KeyLength = defaultDataKeyLength
+	}
+
+	if err := c.KeyLength.Validate(); err != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Cause: err,
+		}
+	}
+
+	// DefaultConfig reads BAO_ADDR and some other optional env variables.
+	config := openbao.DefaultConfig()
+	if config.Error != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Cause: config.Error,
+		}
+	}
+
+	// Address from HCL supersedes BAO_ADDR.
+	if c.Address != "" {
+		config.Address = c.Address
+	}
+
+	client, err := newClient(config, c.Token)
+	if err != nil {
+		return nil, nil, &keyprovider.ErrInvalidConfiguration{
+			Cause: err,
+		}
+	}
+
+	return &keyProvider{
+		svc:       service{client},
+		keyName:   c.KeyName,
+		keyLength: c.KeyLength,
+	}, new(keyMeta), nil
+}
+
+const defaultDataKeyLength DataKeyLength = 32
+
+type DataKeyLength int
+
+func (l DataKeyLength) Validate() error {
+	switch l {
+	case 16, 32, 64:
+		return nil
+	default:
+		return fmt.Errorf("data key length should one of 16, 32 or 64 bytes: got %v", l)
```

Comment:
I'm confused here :-) While these are acceptable values for _Transit_, are these all acceptable values for OpenTofu? I'd imaging nothing in OpenTofu's encryption mechanisms use a 512-bit key and thus 64 should be excluded?

Reply:
We limit OpenBao key provider to the limits of OpenBao itself. Currently, there is no practical sense to use a 64-byte data key, but we plan to extend the list of supported encryption methods.

Reply:
The key provider should still support everything OpenBao supports. The introduction of a new method shouldn't necessitate a change in the key provider because we'll forget to add it.

Reply:
> The key provider should still support everything OpenBao supports. The introduction of a new method shouldn't necessitate a change in the key provider because we'll forget to add it.

Hmm, I'm still confused what happens here. You generate a 512-bit data key and then... Truncate it? Use it verbatim as if it was an AES key (only to get rejected later at AES key construction time)?

Reply:
Yes, it will lead to an error from the AES method. I agree we shouldn't let users make a mistake, but for this exact case, it's more about exposing different building blocks (and yes, some of them could be incompatible) to allow agile configuration.

Reply:
This is meant to be a pluggable architecture. The OpenBao provider shouldn't try to do validation for the AES-GCM method.

Reply:
@janosdebugs Not sure that I or others believe in fully plugable architectures when it comes to cryptography. Often there's subtle issues that can arise. For a modern example, consider the [removed SHA-3 invocation](https://words.filippo.io/dispatches/mlkem768/#bonus-track-using-a-ml-kem-implementation-as-kyber-v3) in the final steps of ML-KEM (no longer providing ciphertext commitment in the shared secret compared to round-3 Kyber). If a hypothetical scheme/protocol relied implicitly on that ciphertext commitment, it would no longer have that with ML-KEM (see: lengthy discussion on IETF list about CMS and whether it needs this -- luckily many protocols don't, but it is why [XWing](https://eprint.iacr.org/2024/039.pdf) was carefully designed).

More concretely to this discussion, consider if a provider only allowed 16 byte keys, but the caller separately had requested AES-256 or ChaCha20 which require 32 bytes? If you zero-pad your key, you end up with an unexpectedly working result, with potentially [weird security properties](https://en.wikipedia.org/wiki/Weak_key). Or, while less unexpected, would you truncate a 64-byte key down to 16 or 32 for AES? Consider too some error cases (e.g., in the future, a [KEM](https://en.wikipedia.org/wiki/Key_encapsulation_mechanism) could be used to effectively generate local data keys without server input, which often has a failure mode of an all-zero shared secret -- [sometimes probabilistically even with fully-valid inputs](https://www.ietf.org/archive/id/draft-connolly-tls-mlkem-key-agreement-01.html#name-discussion)). Would you pass along this all zero key, or err out appropriately? Which layer is responsible for that, the do-AES layer or the get-a-key-from-a-provider layer? &c.

Lastly, I'm not aware of any safe ciphers in widespread use (AES, ChaCha20, or Ascon) that have a 512-bit input key.

I guess, this all to say, its a nice ideal, but tying provider closely to the symmetric algorithm (as a "bundle" perhaps), isn't the worst idea, even if you use common code between them. In that sense, removing the 512-bit recommendation would be considered safer cryptographic design.

Better, removing this option altogether and letting the choice of algorithm dictate the data key size rather than make it explicitly configurable to the user, because invalid states shouldn't be representable. My 2c.